### PR TITLE
Validate API Gateway Response Keys

### PR DESCRIPTION
**Pull request by awood45**
*Issue #, if available: Resolves Issue #522

*Description of changes:* Per my testing, API Gateway has 4 allowable response keys: `statusCode`, `body`, `headers`, and `isBase64Encoded`. None of these keys are required, but any extra keys will result in a 502 error from the service. Currently, the local API Gateway implementation will ignore extra keys, this change will cause those keys to raise an error, matching service behavior.

*Checklist:*

- [N/A] Write Design Document ([Do I need to write a design document?](https://github.com/awslabs/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.rst#design-document))
- [Y] Write unit tests
- [N/A] Write/update functional tests
- [Y] Write/update integration tests
- [Y] `make pr` passes
- [N/A] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: samcli/local/apigw/local_apigw_service.py [hunk 1]
```diff
@@ -230,6 +235,17 @@ def _parse_lambda_output(lambda_output, binary_types, flask_request):
...
+    @staticmethod
+    def _invalid_apig_response(output):
+        allowable = [
+            "statusCode",
+            "body",
+            "headers",
+            "isBase64Encoded"
+        ]
+        invalid_keys = set(output) - set(allowable)
+        return invalid_keys != set()
```
Comment (sanathkr) [inline samcli/local/apigw/local_apigw_service.py]: `return bool(invalid_keys)` is sufficient! In Python empty list/set is considered Falsy. The `!=` syntax is not too Pythonic ;)

File: samcli/local/apigw/local_apigw_service.py [hunk 2]
```diff
@@ -230,6 +235,17 @@ def _parse_lambda_output(lambda_output, binary_types, flask_request):
...
 
+    @staticmethod
+    def _invalid_apig_response(output):
+        allowable = [
+            "statusCode",
+            "body",
+            "headers",
+            "isBase64Encoded"
+        ]
+        invalid_keys = set(output) - set(allowable)
```
Comment (sanathkr) [inline samcli/local/apigw/local_apigw_service.py]: nit: Is `output` dict? If so explicitly grab the keys using `output.keys()`.

Comment (awood45) [reply samcli/local/apigw/local_apigw_service.py]: Updated, and tests still passing locally.

File: samcli/local/apigw/local_apigw_service.py (hunk 1 above)
Comment (awood45) [reply samcli/local/apigw/local_apigw_service.py]: Same story, tests still pass after this change, just merged.

File: tests/integration/local/start_api/test_start_api.py [hunk 3]
```diff
@@ -86,6 +86,12 @@ def test_invalid_response_from_lambda(self):
         self.assertEquals(response.status_code, 502)
         self.assertEquals(response.json(), {"message": "Internal server error"})
 
+    def test_invalid_hash_response_from_lambda(self):
```
Comment (TheSriram) [inline tests/integration/local/start_api/test_start_api.py]: nit: json_response?

File: tests/integration/testdata/start_api/main.py [hunk 4]
```diff
@@ -63,6 +63,10 @@ def invalid_response_returned(event, context):
     return "This is invalid"
 
 
+def invalid_hash_response(event, context):
```
Comment (TheSriram) [inline tests/integration/testdata/start_api/main.py]: nit: maybe just json response?

File: samcli/local/apigw/local_apigw_service.py [hunk 5]
```diff
@@ -230,6 +235,17 @@ def _parse_lambda_output(lambda_output, binary_types, flask_request):
 
         return status_code, headers, body
 
+    @staticmethod
+    def _invalid_apig_response(output):
+        allowable = [
```
Comment (TheSriram) [inline samcli/local/apigw/local_apigw_service.py]: you could make allowable a set to begin with.
`
allowable = {...}
`

File: tests/integration/local/start_api/test_start_api.py (hunk 3 above)
Comment (awood45) [reply tests/integration/local/start_api/test_start_api.py]: Sure

File: samcli/local/apigw/local_apigw_service.py (hunk 5 above)
Comment (awood45) [reply samcli/local/apigw/local_apigw_service.py]: Changed, though I may have to change further as the set subtraction operator doesn't seem to be supported in Python 2.7

File: samcli/local/apigw/local_apigw_service.py [hunk 6]
```diff
@@ -220,6 +220,11 @@ def _parse_lambda_output(lambda_output, binary_types, flask_request):
             LOG.error(message)
             raise TypeError(message)
 
+        # API Gateway only accepts statusCode, body, headers, and isBase64Encoded in
+        # a response shape.
+        if LocalApigwService._invalid_apig_response(json_output):
+            raise ValueError
```
[the remark this answers is not in the document]
Comment (awood45) [reply samcli/local/apigw/local_apigw_service.py]: Done

File: samcli/local/apigw/local_apigw_service.py (hunk 5 above)
Comment (awood45) [reply samcli/local/apigw/local_apigw_service.py]: I made this change, just had to add the `set()` casting explicitly to the keys.

**Comment by awood45** [comment]
Python 2.7 incompatibility is fixed, so this is ready for review again.

File: samcli/local/apigw/local_apigw_service.py [hunk 7]
```diff
@@ -220,6 +220,12 @@ def _parse_lambda_output(lambda_output, binary_types, flask_request):
             LOG.error(message)
             raise TypeError(message)
 
+        # API Gateway only accepts statusCode, body, headers, and isBase64Encoded in
+        # a response shape.
+        if LocalApigwService._invalid_apig_response(json_output):
+            msg = "Invalid Response: " + str(json_output)
```
Comment (TheSriram) [inline samcli/local/apigw/local_apigw_service.py]: do we need to be more explicit in the response about accepted keys?

Comment (TheSriram) [reply samcli/local/apigw/local_apigw_service.py]: we can possibly move log statement up here just before throwing an exception.

File: samcli/local/apigw/local_apigw_service.py [hunk 8]
```diff
@@ -230,6 +236,20 @@ def _parse_lambda_output(lambda_output, binary_types, flask_request):
...
 
+    @staticmethod
+    def _invalid_apig_response(output):
+        allowable = {
+            "statusCode",
+            "body",
+            "headers",
+            "isBase64Encoded"
+        }
+        # In Python 2.7, need to explicitly make the Dictionary keys into a set
```
Comment (TheSriram) [inline samcli/local/apigw/local_apigw_service.py]: Good to know!

File: samcli/local/apigw/local_apigw_service.py (hunk 7 above)
Comment (awood45) [reply samcli/local/apigw/local_apigw_service.py]: To do this I'd probably want to actually return the invalid keys and then perform the boolean if at this top level. Should we move to that approach?

Comment (TheSriram) [reply samcli/local/apigw/local_apigw_service.py]: Yeah that sounds good, provides more context as well.

Comment (awood45) [reply samcli/local/apigw/local_apigw_service.py]: Addressed in most recent change.

**Pull request merged**